src/__tests__/cacheDirectory.test.ts

```diff
@@ -0,0 +1,176 @@
+import { promises as fs } from 'fs';
+import { v4 as uuid } from 'uuid';
+import { vol } from 'memfs';
+
+import {
+  getDefaultCacheDirectory,
+  writeJsonToPath,
+  symlink,
+} from '../cacheDirectory';
+
+jest.mock('fs'); // applies manual mock which uses memfs
+
+afterEach(() => {
+  jest.clearAllMocks();
+  vol.reset(); // clear out memory
+});
+
+describe('getDefaultCacheDirectory', () => {
+  test('should utilize the current directory for building the default cache directory', () => {
+    const mockCwdResult = `/${uuid()}`;
+    jest.spyOn(process, 'cwd').mockReturnValue(mockCwdResult);
+
+    expect(getDefaultCacheDirectory()).toEqual(
+      `${mockCwdResult}/.j1-integration`,
+    );
+  });
+});
+
+describe('writeJsonToPath', () => {
+  test('should pretty write json to the specified file', async () => {
+    const json = { test: '123' };
+
+    const directory = '/';
+    const filename = `${uuid()}.json`;
+
+    await writeJsonToPath({
+      cacheDirectory: directory,
+      path: filename,
+      data: json,
+    });
+
+    const writtenData = await fs.readFile(`${directory}/${filename}`, 'utf8');
+    expect(writtenData).toEqual(JSON.stringify(json, null, 2));
+  });
+
+  test('should recursively create directories prior to writing', async () => {
+    const json = { woah: 'json' };
+
+    const directory = '/';
+    const filename = `test/dir/that/does/not/already/exist/${uuid()}.json`;
+
+    const mkdirSpy = jest.spyOn(fs, 'mkdir');
+    const writeFileSpy = jest.spyOn(fs, 'writeFile');
+
+    await writeJsonToPath({
+      cacheDirectory: directory,
+      path: filename,
+      data: json,
+    });
+
+    const writtenData = await fs.readFile(`${directory}/${filename}`, 'utf8');
+    expect(writtenData).toEqual(JSON.stringify(json, null, 2));
+
+    expect(mkdirSpy).toHaveBeenCalledTimes(1);
+    expect(mkdirSpy).toHaveBeenCalledWith(
+      '/test/dir/that/does/not/already/exist',
+      {
+        recursive: true,
+      },
+    );
+
+    // not pretty but wanted to enforce that mkdir was being called before
+    // write file
+    //
+    // ref: https://github.com/facebook/jest/issues/4402#issuecomment-534516219
+    const mkdirOrder = mkdirSpy.mock.invocationCallOrder[0];
+    const writeFileOrder = writeFileSpy.mock.invocationCallOrder[0];
+    expect(mkdirOrder).toBeLessThan(writeFileOrder);
```

Comment:
Does `memfs` not raise an error if an attempt is made to write to a file path that references a directory(ies) that does not exist? Or to say it another way, a normal file system would fail to write the file if the directory doesn't exist, so when the file is written/read in the test, you can know that `mkdir` must have been called.

Reply:
It does fail if the directory doesn't exist like a regular file system. I suppose the assertion here is a bit overkill.